src/cdk/table/array-data-source.ts

```diff
@@ -0,0 +1,43 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DataSource} from './data-source';
+import {BehaviorSubject} from 'rxjs/BehaviorSubject';
+
+/**
+ * Array-based data source that can be used to connect to a CollectionViewer. Will trigger an
+ * update to the CollectionViewer whenever a new data array is set.
+ */
+export class ArrayDataSource<T> implements DataSource<T> {
+  /** Stream that emits when a new data array is set on the data source. */
+  private _data: BehaviorSubject<T[]>;
+  set data(data: T[]) { this._data.next(data); }
+  get data() { return this._data.value; }
+
+  constructor(initialData: T[] = []) {
+    this._data = new BehaviorSubject<T[]>(initialData);
+  }
+
+  /**
+   * Triggers an update to the CollectionViewer. Should be used when elements are added, removed,
+   * or moved in the data array and should be reflected in the CollectionViewer.
+   */
+  update() { this._data.next(this.data); }
```

Comment:
I'm just a little worried about adding this caveat (even though I understand it's just how array references work). And I expect it will raise a bunch of issues like md-select's compare by reference has done.

Maybe one of these could work? The advantage (up for argument 😏) is that `update()` is required, so no matter what the user does with their array and `arrayDataSource.data`, they can expect that nothing will change until they call `update()`.

1. Users must supply the new data to `update()`
```ts
  private _data: BehaviorSubject<T[]>;
  get data() { return this._data.value; }

  constructor(initialData: T[] = []) {
    this._data = new BehaviorSubject<T[]>(initialData);
  }

  update(newData: T[]) { this._data.next(newData); }
```
2. Users can set/manipulate `data` all they want, but to render a change, they **must** call `update()`
```ts
  private _data: BehaviorSubject<T[]>;
  public data: T[];

  constructor(initialData: T[] = []) {
    this._data = new BehaviorSubject<T[]>(initialData);
  }

  update() { this._data.next(this.data); }
```